File: coco/src/main/java/coco/ep/i2/Power.java
package coco.ep.i2;

import coco.ep.i1.Exp;

public interface Power<FT> extends Exp<FT>, Factory<FT> {

    Exp<FT> getLeft();

    Exp<FT> getRight();

    default String prettyp() {
        return "(" + convert(this.getLeft()).prettyp() + "^" + convert(this.getRight()).prettyp() + ")";
    }

    default coco.ep.Exp<FT> multby(coco.ep.Exp<FT> other) {
        double leftEval = convert(this.getLeft()).eval();
        double added = Math.log(convert(other).eval()) / Math.log(leftEval);
        return this.power(this.getLeft(), this.add(this.getRight(), this.lit(added)));
    }

    default Double eval() {
        return Math.pow(convert(this.getLeft()).eval(), convert(this.getRight()).eval());
    }
}

File: coco_csharp/CoCo/EP/I2/Power.cs
using System;

namespace EP.I2 {
    public interface Power<FT> : EP.I1.Exp<FT>, Factory<FT> {

        EP.Exp<FT> getLeft();

        EP.Exp<FT> getRight();

        double EP.M0.Exp<FT>.eval() {
            return Math.Pow(convert(getLeft()).eval(), convert(getRight()).eval());
        }

        EP.Exp<FT> EP.I1.Exp<FT>.multby(EP.Exp<FT> other) {
            double leftEval = convert(getLeft()).eval();
            double added = Math.Log(convert(other).eval()) / Math.Log(leftEval);
            return power(getLeft(), add(getRight(), lit(added)));
        }

        string EP.M2.Exp<FT>.prettyp() { return "(" + convert(getLeft()).prettyp() + "^" + convert(getRight()).prettyp() + ")"; }
    }
}